### PR TITLE
Local backend validation#433

### PR DESCRIPTION
**Pull request by gauravs90**
Solves #433. It goes hand in hand with the PR in cromwell-backends (https://github.com/broadinstitute/cromwell-backend/pull/26)

In addition, have cleaned up the WorkflowDescriptor a bit.

### CODE REVIEW

**Comment by gauravs90** [comment]
Sorry @mcovarr  and @francares , it had to be you guys again in-order to relate with the other PR in cromwell-backends :D

File: src/main/scala/cromwell/engine/workflow/ValidateActor.scala [hunk 1]
```diff
@@ -10,65 +15,181 @@ import wdl4s._
...
+    * this workflow.
+    * Later on, this method can be changed to a 'per-call-basis' acceptance test by the backends,
+    * i.e. it may return a Map of [Call -> ExecutableBackends]
+    *
+    * @param namespace     workflow with the namespaces already resolved
+    * @param wfInputs      Already coerced inputs
+    * @param wfOptionsJson workflow options
+    * @return List of BackendConfigurationEntry objects, representing which backends can accept this WF
+    */
+  private def executableBackendsForWf(namespace: NamespaceWithWorkflow,
```
Comment (jainh) [inline src/main/scala/cromwell/engine/workflow/ValidateActor.scala]: Name of the method is not intuitive, use proper semantics  as per your comment. It should be something like validationOfExecutableBackends or(validationOfPluggableBackends). Second break this method in two parts 1) proper validation for backends that have validation class 2) another method bringing will execute backends.

File: src/main/scala/cromwell/engine/workflow/ValidateActor.scala [hunk 2]
```diff
@@ -10,65 +15,181 @@ import wdl4s._
...
+    val (mayExecute, willExecute) = listOfPluggedBackends.partition(_.validationClass.isDefined)
+
+    if (willExecute.nonEmpty) logger.info(s"Backends with no validation requirements: $willExecute")
+
+    val backendEntryValidatorActorPair: Map[BackendConfigurationEntry, ActorRef] = mayExecute map { entry =>
+      entry -> DefaultBackendFactory.getBackendActorFor(entry.validationClass.get, context.system, namespace, wfInputs, wfOptionsJson)
+    } toMap
+
+    // The implicit timeout required for Ask below comes from `CromwellActor`
+    val mapOfEntryToFutureValidationRes: Map[BackendConfigurationEntry, Future[ValidationResult]] = backendEntryValidatorActorPair mapValues { validateActor =>
```
Comment (jainh) [inline src/main/scala/cromwell/engine/workflow/ValidateActor.scala]: looks like these two vals (backendEntryValidatorActorPair,mapOfEntryToFutureValidationRes) can be combined like : f1 andThen f2

File: src/main/scala/cromwell/engine/WorkflowDescriptor.scala [hunk 3]
```diff
@@ -152,113 +151,48 @@ case class WorkflowDescriptor(id: WorkflowId,
...
-    Try(namespace.workflow.calls foreach { x => x.task.runtimeAttributes }) match {
-      case scala.util.Success(_) => ().successNel
-      case scala.util.Failure(e) => s"Workflow $id contains bad runtime attributes: ${e.getMessage}".failureNel
+  //TODO: Move it to the ValidateActor part, can't move now since this has engine functions (i.e. Backend) dependency.
+  private def validateDeclarations(id: WorkflowId,
+                                   namespace: NamespaceWithWorkflow,
+                                   coercedInputs: WorkflowCoercedInputs,
+                                   engineFunctions: WorkflowEngineFunctions): WorkflowCoercedInputs = {
+    namespace.staticDeclarationsRecursive(coercedInputs, engineFunctions) match {
+      case Success(d) => d
```
Comment (jainh) [inline src/main/scala/cromwell/engine/WorkflowDescriptor.scala]: what is d? name should be proper.

File: src/main/scala/cromwell/engine/workflow/ValidateActor.scala [hunk 4]
```diff
@@ -10,65 +15,181 @@ import wdl4s._
...
+      }
+    }.map(_.toMap)
+
+    //Just logging
+    futureMapOfEntryToValidationRes foreach {
+      _ foreach { case (k, v) => logger.info(s"For Backend [${k.name}], Result: [$v]") }
+    }
+
+    //Map which does not have any backends which failed validation
+    val validatedBackends: Future[Map[BackendConfigurationEntry, ValidationResult]] = futureMapOfEntryToValidationRes map {
```
Comment (jainh) [inline src/main/scala/cromwell/engine/workflow/ValidateActor.scala]: same as above.

File: src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala [hunk 5]
```diff
@@ -166,9 +169,22 @@ class WorkflowManagerActor() extends LoggingFSM[WorkflowManagerState, WorkflowMa
...
 
   when (Running) {
-    case Event(SubmitWorkflow(source), _) =>
-      val updatedData = submitWorkflow(source, replyTo = Option(sender), id = None)
-      stay() using updatedData
+    case Event(ValidateAndSubmitWorkflow(source), _) =>
+      import akka.pattern.ask
+      val requester = sender()
+      val validateActor = context.actorOf(ValidateActor.props(source))
+      val validationResFut = (validateActor ? RequestValidation).mapTo[ValidateActorMessage]
```
Comment (jainh) [inline src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala]: do we need to use ask here ? I then fire and forget should work?

**Pull request closed without merging**